gluten-core/src/main/scala/org/apache/spark/util/TaskResources.scala

```diff
@@ -164,44 +162,39 @@ object TaskResources extends TaskListener with Logging {
 // thread safe
 class TaskResourceRegistry extends Logging {
   private val sharedUsage = new SimpleMemoryUsageRecorder()
-  private val resources = new java.util.LinkedHashMap[String, TaskResource]()
-  private val resourcesPriorityMapping =
-    new java.util.HashMap[Long, java.util.List[TaskResource]]()
+  private val resources = new util.HashMap[String, TaskResource]()
+  type TaskResourceWithOrdering = (TaskResource, Int)
+  // 0 is lowest, Int.MaxValue is highest
+  private val resourcesPriorityQueue =
+    new PriorityQueue[TaskResourceWithOrdering](new Comparator[TaskResourceWithOrdering]() {
+      override def compare(t1: TaskResourceWithOrdering, t2: TaskResourceWithOrdering): Int = {
+        val diff = t2._1.priority() - t1._1.priority()
+        if (diff == 0) {
+          // If the task resource has same priority, we should follow the LIFO
+          t2._2 - t1._2
+        } else {
+          diff
+        }
+      }
+    })
 
   private def addResource0(id: String, resource: TaskResource): Unit = synchronized {
     resources.put(id, resource)
-    if (!resourcesPriorityMapping.containsKey(resource.priority())) {
-      resourcesPriorityMapping.put(resource.priority(), new util.ArrayList[TaskResource]())
-    }
-    val list = resourcesPriorityMapping.get(resource.priority())
-    list.add(resource)
+    resourcesPriorityQueue.add((resource, resourcesPriorityQueue.size()))
```

Comment:
This is buggy. The queue size would not be predictable for the same priority since it is the size for all priorities. If no other way you can find, probably dual nesting is not a bad choice.

Reply:
it always monotonically increasing as we never remove the element, so I think should be fine ?

Reply:
I understand your point however the assumption relies on the fact that one should program carefully within this principle: never call `poll()` unless during task ends. 

It could be an option to have a accumulator to generated increasing integer IDs? That may align with your original purpose better I think.

Reply:
I'm fine to add one more accumulator, addressed